File: examples/payment-event-receiver/python/receiver.py
#!/usr/bin/env python3
"""
BC Registries – Partner payment-event webhook receiver (Python).

Uses the stdlib http.server — no web framework required.
JWT verification uses google-auth against Google's OIDC endpoint.

Environment variables:
    AUDIENCE        Audience the JWT must carry — set to this server's public URL.
    PUBSUB_SA_EMAIL Service account email Pub/Sub signs tokens as.

Run:
    pip install -r requirements.txt
    AUDIENCE=https://... python receiver.py
"""

import json
import logging
import os
from http.server import BaseHTTPRequestHandler, HTTPServer

import google.auth.transport.requests
import google.oauth2.id_token

logging.basicConfig(level=logging.INFO, format="%(asctime)s %(levelname)s %(message)s")
log = logging.getLogger(__name__)

AUDIENCE = os.environ["AUDIENCE"]
EXPECTED_SA_EMAIL = os.environ["PUBSUB_SA_EMAIL"]

_http_request = google.auth.transport.requests.Request()


def verify_jwt(token: str) -> dict:
    """Verify a Pub/Sub OIDC JWT and return its decoded claims.

    GCP Pub/Sub signs every push request with an OIDC token issued for the
    configured delivery service account. This function:
      1. Fetches Google's public keys (cached internally by google-auth).
      2. Verifies the RS256 signature.
      3. Confirms the audience matches AUDIENCE (this server's URL).
      4. Confirms the token has not expired.

    Raises ValueError (or google.auth.exceptions.TransportError) if any
    check fails. The caller should treat any exception as an invalid token.

    Returns a dict of JWT claims, e.g. {"email": "...", "sub": "...", ...}.
    """
    return google.oauth2.id_token.verify_oauth2_token(token, _http_request, audience=AUDIENCE)


class Handler(BaseHTTPRequestHandler):
    def do_POST(self):
        if self.path != "/pay-events":
            self._send(404)
            return

        auth = self.headers.get("Authorization", "")
        if not auth.startswith("Bearer "):
            self._send(401, {"error": "missing bearer token"})
            return

        try:
            claims = verify_jwt(auth.removeprefix("Bearer "))
        except Exception as exc:
            log.warning("JWT verification failed: %s", exc)
            self._send(401, {"error": "invalid token"})
            return

        if claims.get("email") != EXPECTED_SA_EMAIL:
            log.warning("unexpected caller: %s", claims.get("email"))
            self._send(403, {"error": "unexpected service account"})
            return

        length = int(self.headers.get("Content-Length", 0))
        raw = self.rfile.read(length).decode()
        try:
            body = json.loads(raw)
        except json.JSONDecodeError:
            body = raw

        log.info(
            "event  message_id=%s  publish_time=%s\n%s",
            self.headers.get("X-Goog-Message-Id", "-"),
            self.headers.get("X-Goog-Publish-Time", "-"),
            json.dumps(body, indent=2) if isinstance(body, dict) else body,
        )
        self._send(204)

    def _send(self, status: int, body: dict | None = None):
        self.send_response(status)
        if body:
            payload = json.dumps(body).encode()
            self.send_header("Content-Type", "application/json")
            self.send_header("Content-Length", len(payload))
            self.end_headers()
            self.wfile.write(payload)
        else:
            self.end_headers()

    def log_message(self, *_):
        pass  # silence default access log; we handle logging ourselves


if __name__ == "__main__":
    server = HTTPServer(("", 8090), Handler)
    log.info("listening on :8090")
    server.serve_forever()

File: examples/payment-event-receiver/nodejs/receiver.js
'use strict';
/**
 * BC Registries – Partner payment-event webhook receiver (Node.js).
 *
 * Uses the stdlib http module — no web framework required.
 * JWT verification uses google-auth-library against Google's OIDC endpoint.
 *
 * Environment variables:
 *   AUDIENCE        Audience the JWT must carry — set to this server's public URL.
 *   PUBSUB_SA_EMAIL Service account email Pub/Sub signs tokens as.
 *
 * Run:
 *   npm install
 *   AUDIENCE=https://... node receiver.js
 */

const http = require('http');
const { OAuth2Client } = require('google-auth-library');

const AUDIENCE = process.env.AUDIENCE;
const EXPECTED_SA_EMAIL = process.env.PUBSUB_SA_EMAIL;

if (!AUDIENCE) { console.error('AUDIENCE is required'); process.exit(1); }
if (!EXPECTED_SA_EMAIL) { console.error('PUBSUB_SA_EMAIL is required'); process.exit(1); }

const _defaultClient = new OAuth2Client();

/**
 * Verify a Pub/Sub OIDC JWT and return its decoded claims.
 *
 * GCP Pub/Sub signs every push request with an OIDC token issued for the
 * configured delivery service account. This function:
 *   1. Fetches Google's public keys (cached internally by google-auth-library).
 *   2. Verifies the RS256 signature.
 *   3. Confirms the audience matches AUDIENCE (this server's URL).
 *   4. Confirms the token has not expired.
 *
 * Throws if any check fails — the caller should treat any error as an invalid token.
 * Returns the payload object, e.g. { email: '...', sub: '...', ... }.
 *
 * @param {string} token - Raw JWT string from the Authorization: Bearer header.
 * @param {OAuth2Client} [authClient] - Override for testing; defaults to a shared client.
 */
async function verifyJwt(token, authClient = _defaultClient) {
    const ticket = await authClient.verifyIdToken({ idToken: token, audience: AUDIENCE });
    return ticket.getPayload();
}

function readBody(req) {
    return new Promise((resolve, reject) => {
        const chunks = [];
        req.on('data', c => chunks.push(c));
        req.on('end', () => resolve(Buffer.concat(chunks).toString('utf8')));
        req.on('error', reject);
    });
}

function send(res, status, body) {
    if (body) {
        const payload = JSON.stringify(body);
        res.writeHead(status, { 'Content-Type': 'application/json', 'Content-Length': Buffer.byteLength(payload) });
        res.end(payload);
    } else {
        res.writeHead(status);
        res.end();
    }
}

const server = http.createServer(async (req, res) => {
    if (req.method !== 'POST' || req.url !== '/pay-events') {
        return send(res, 404);
    }

    const auth = req.headers['authorization'] ?? '';
    if (!auth.startsWith('Bearer ')) {
        return send(res, 401, { error: 'missing bearer token' });
    }

    let claims;
    try {
        claims = await verifyJwt(auth.slice(7));
    } catch (err) {
        console.warn('JWT verification failed:', err.message);
        return send(res, 401, { error: 'invalid token' });
    }

    if (claims.email !== EXPECTED_SA_EMAIL) {
        console.warn('unexpected caller:', claims.email);
        return send(res, 403, { error: 'unexpected service account' });
    }

    const raw = await readBody(req);
    let body;
    try { body = JSON.parse(raw); } catch { body = raw; }

    console.log(
        `event  message_id=${req.headers['x-goog-message-id'] ?? '-'}` +
        `  publish_time=${req.headers['x-goog-publish-time'] ?? '-'}`
    );
    console.log(typeof body === 'object' ? JSON.stringify(body, null, 2) : body);

    send(res, 204);
});

// Guard lets test files require this module without starting the server.
if (require.main === module) {
    server.listen(8090, () => console.log('listening on :8090'));
}

module.exports = { verifyJwt };
